### PR TITLE
Allow for 'DEFERRED' and 'ATTRIBUTE' forms of unknown array extent (594)

### PR DESCRIPTION
**Pull request by arporter**
I got impatient so went ahead and implemented it :-) We can continue the debate over naming as part of the review process...

### CODE REVIEW

**Comment by arporter** [comment]
Just waiting on Travis for coverage...

**Comment by arporter** [comment]
Travis is happy - ready for first review.

File: src/psyclone/tests/psyir/backend/c_test.py [hunk 1]
```diff
@@ -70,14 +70,16 @@ def test_cw_gen_declaration():
...
     # Array argument
-    symbol = DataSymbol("dummy2", "real", shape=[2, None, 2],
+    symbol = DataSymbol("dummy2", "real",
+                        shape=[2, DataSymbol.Extent.ATTRIBUTE, 2],
                         interface=ArgumentInterface(
                             ArgumentInterface.Access.READ))
     result = cwriter.gen_declaration(symbol)
     assert result == "double * restrict dummy2"
 
     # Array with unknown intent
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/tests/psyir/backend/c_test.py]: Fixed although I didn't capitalise as I don't think that's right.

File: src/psyclone/psyir/symbols/datasymbol.py [hunk 2]
```diff
@@ -100,6 +104,16 @@ class Precision(Enum):
         SINGLE = 1
         DOUBLE = 2
 
+    class Extent(Enum):
+        '''
+        Arrays may have one or more dimensions of unknown extent.
+        Where the extent must exist and is accessible via the run-time system
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/symbols/datasymbol.py]: Yes.

File: src/psyclone/psyir/symbols/datasymbol.py [hunk 3]
```diff
@@ -100,6 +104,16 @@ class Precision(Enum):
         SINGLE = 1
         DOUBLE = 2
 
+    class Extent(Enum):
+        '''
+        Arrays may have one or more dimensions of unknown extent.
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/symbols/datasymbol.py]: Yes, your suggestion is better. I was thinking along the lines of "An Enumeration class for..." but strictly it is an enumeration of options so I've changed to the latter.

File: src/psyclone/psyir/frontend/fparser2.py [hunk 4]
```diff
@@ -635,6 +638,21 @@ def process_declarations(self, parent, nodes, arg_list):
...
+                if allocatable and not entity_shape:
+                    # We have an allocatable attribute on something that we
+                    # don't recognise as an array - this is not supported.
+                    raise NotImplementedError(
+                        "Could not process {0}. The 'allocatable' attribute is"
+                        " only supported on array declarations.".format(
+                            decl.items))
+
+                for idx, extent in enumerate(entity_shape):
+                    if extent is None:
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/frontend/fparser2.py]: That's true. Hopefully the parser should spot that but just in case it doesn't I've added a check.

[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/frontend/fparser2.py]: fparser doesn't spot it so I've created https://github.com/stfc/fparser/issues/229 for it.

**Comment by arporter** [comment]
On discussing with @sergisiso I realised I hadn't written any documentation (although, in my defence, the UG contains an auto-generated description of DataSymbol). I've therefore added a short section describing the use of `shape` when defining an array.
This is now ready for next review.

File: doc/user_guide/psyir.rst [hunk 5]
```diff
@@ -122,15 +122,43 @@ The ``SymbolTable`` has the following interface:
...
 
 - .. autoclass:: psyclone.psyir.symbols.DataSymbol
 
+As described in the ``DataSymbol`` constructor, the ``shape``
+specification not only describes whether or not the ``Symbol`` is an
+array but also, if it is an array, captures what is known about how the extent
+(size) of each dimension is managed. This is necessary to distinguish
+between four cases:
+
+.. tabularcolumns:: |p{9cm}|L|
```
Comment (sergisiso) [inline doc/user_guide/psyir.rst]: Not part of this PR, but the table does not break into paragraphs and uses long lines instead (at least in html).

Also, do the description sentences need punctuation?

File: doc/user_guide/psyir.rst [hunk 6]
```diff
@@ -122,15 +122,43 @@ The ``SymbolTable`` has the following interface:
...
+|An array has a definite extent which is not| ``DataSymbol.Extent.ATTRIBUTE``|
+|known at compile time but can be queried   |                                |
+|at runtime                                 |                                |
++-------------------------------------------+--------------------------------+
+|It is not known whether an array has memory| ``DataSymbol.Extent.DEFERRED`` |
+|allocated to it in the current scoping unit|                                |
++-------------------------------------------+--------------------------------+
+
+(The distinction between the last two cases is necessary to allow the
+Fortran backend to the PSyIR to create correct code for a subroutine
```
Comment (sergisiso) [inline doc/user_guide/psyir.rst]: Maybe "..to allow the PSyIR Fortran backend to create.." or "..the PSyIR to Fotran backend.."

File: src/psyclone/tests/psyir/backend/fortran_test.py [hunk 7]
```diff
@@ -267,19 +268,30 @@ def test_fw_gen_vardecl(fort_writer):
...
 
+    # Allocatable array
+    symbol = DataSymbol("dummy2", "real", shape=[DataSymbol.Extent.DEFERRED,
+                                                 DataSymbol.Extent.DEFERRED],
+                        interface=ArgumentInterface(
+                            ArgumentInterface.Access.READWRITE))
+    result = fort_writer.gen_vardecl(symbol)
+    assert result == \
+        "real, allocatable, dimension(:,:), intent(inout) :: dummy2\n"
+
```
Comment (sergisiso) [inline src/psyclone/tests/psyir/backend/fortran_test.py]: Since it is 'at least 1 DEFERRED' to be an allocatable maybe it will be good to have a second test array with something like "shape=[2, DataSymbol.Extent.DEFERRED, 2]"?

File: doc/user_guide/psyir.rst (hunk 6 above)
Comment (arporter) [reply doc/user_guide/psyir.rst]: Yes, that's better.

File: doc/user_guide/psyir.rst (hunk 5 above)
Comment (arporter) [reply doc/user_guide/psyir.rst]: Yes, I'm happy to leave long lines in html (although I know Joerg doesn't like it). I've added punctuation.

File: src/psyclone/tests/psyir/backend/fortran_test.py (hunk 7 above)
Comment (arporter) [reply src/psyclone/tests/psyir/backend/fortran_test.py]: The 'at least 1 DEFERRED' was to cover languages other than Fortran (an allocatable array in Fortran must have every dimension specified as ':'). I've added a check for this in the visitor plus associated test.

Comment (arporter) [reply src/psyclone/tests/psyir/backend/fortran_test.py]: Actually, the situation is more complicated than that - arrays that are dummy arguments may be of 'assumed size' where the extents of one or more of the outer ranks of an array are unspecified, e.g.:

    integer, intent(in) :: my_array(3,:)

Comment (arporter) [reply src/psyclone/tests/psyir/backend/fortran_test.py]: An array that is a dummy argument may be 'assumed shape' in which case all ranks are specified with ':' and the extents are taken from the actual argument. This actual array *need not* be allocatable.

Comment (arporter) [reply src/psyclone/tests/psyir/backend/fortran_test.py]: Happily, in both of those cases, the frontend will correctly store the unspecified extents as 'ATTRIBUTE' because there's no 'allocatable' attribute on the declaration.

Comment (arporter) [reply src/psyclone/tests/psyir/backend/fortran_test.py]: I've added an additional check (in the backend) that we are about to generate a valid assumed-size declaration.

**Comment by arporter** [comment]
Ready for next review.
